// nativescript-socket.io/socket.ios.js
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var common_1 = require("./common");
exports.enableDebug = common_1.enableDebug;
exports.disableDebug = common_1.disableDebug;
var helpers_1 = require("./helpers");
SocketIOClient;
SocketAckEmitter;
OnAckCallback;
SocketEngine;
var NAMESPACE_REGEXP = /^https?\:\/\/[^\/]*(\/.*)$/i;
var Socket = (function (_super) {
    __extends(Socket, _super);
    function Socket(uri, options) {
        if (options === void 0) { options = {}; }
        _super.call(this);
        var _options = {};
        if (NAMESPACE_REGEXP.test(uri)) {
            _options.nsp = uri.match(NAMESPACE_REGEXP)[1] || '/';
        }
        else {
            _options.nsp = '/';
        }
        if (options.query) {
            _options.connectParams = {};
            if (typeof options.query === 'string') {
                options.query.split('&').forEach(function (pair) {
                    pair = pair.split('=').map(decodeURIComponent);
                    _options.connectParams[pair[0]] = pair[1];
                });
            }
            else {
                Object.keys(options.query).forEach(function (key) {
                    _options.connectParams[key] = String(options.query[key]);
                });
            }
        }
        if (options.ios) {
            Object.keys(options.ios).forEach(function (prop) {
                _options[prop] = options.ios[prop];
            });
        }
        this.ios = SocketIOClient.alloc().initWithSocketURLConfig(NSURL.URLWithString(uri), _options);
    }
    Object.defineProperty(Socket.prototype, "connected", {
        get: function () {
            return this.ios && this.ios.engine.connected;
        },
        enumerable: true,
        configurable: true
    });
    Socket.prototype.connect = function () {
        this.ios.connect();
    };
    Socket.prototype.disconnect = function () {
        this.ios.disconnect();
    };
    Socket.prototype.on = function (event, callback) {
        var listener = function (data, ack) {
            var payload = helpers_1.deserialize(data);
            common_1.debug('on', event, payload, ack && ack.expected ? 'ack' : '');
            if (ack && ack.expected) {
                var _ack = function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i - 0] = arguments[_i];
                    }
                    common_1.debug('on', event, 'ack', args);
                    args = args.map(helpers_1.serialize);
                    ack.with(args);
                };
                payload.push(_ack);
            }
            callback.apply(null, payload);
        };
        var listenerId = this.ios.onCallback(event, listener);
        this._listeners.set(callback, listenerId);
        return this;
    };
    Socket.prototype.once = function (event, callback) {
        var listener = function (data, ack) {
            var payload = helpers_1.deserialize(data);
            common_1.debug('once', event, payload, ack && ack.expected ? 'ack' : '');
            if (ack && ack.expected) {
                var _ack = function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i - 0] = arguments[_i];
                    }
                    common_1.debug('once', event, 'ack', args);
                    args = args.map(helpers_1.serialize);
                    ack.with(args);
                };
                payload.push(_ack);
            }
            callback.apply(null, payload);
        };
        var listenerId = this.ios.onceCallback(event, listener);
        this._listeners.set(callback, listenerId);
        return this;
    };
    Socket.prototype.off = function (event, callback) {
        common_1.debug('off', event, callback);
        if (callback) {
            var listenerId = this._listeners.get(callback);
            if (listenerId) {
                this.ios.offWithId(listenerId);
                this._listeners.delete(callback);
            }
        }
        else {
            this.ios.off(event);
        }
        return this;
    };
    Socket.prototype.emit = function (event) {
        var payload = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            payload[_i - 1] = arguments[_i];
        }
        var ack = payload.pop();
        if (typeof ack === 'undefined') {
            ack = null;
        }
        else if (typeof ack !== 'function') {
            payload.push(ack);
            ack = null;
        }
        common_1.debug('emit', event, payload, ack ? 'ack' : '');
        payload = payload.map(helpers_1.serialize);
        if (ack) {
            var _ack = function (args) {
                args = helpers_1.deserialize(args);
                common_1.debug('emit', event, 'ack', args);
                ack.apply(null, args);
            };
            this.ios.emitWithAckWith(event, payload).timingOutAfterCallback(0, _ack);
        }
        else {
            this.ios.emitWith(event, payload);
        }
        return this;
    };
    return Socket;
}(common_1.SocketBase));
exports.Socket = Socket;
function connect(uri, options) {
    var socket = new Socket(uri, options || {});
    socket.connect();
    return socket;
}
exports.connect = connect;
